vm_insnhelper.c

```diff
@@ -6386,33 +6386,41 @@ vm_inlined_ic_hit_p(VALUE flags, VALUE value, const rb_cref_t *ic_cref, const VA
 }
 
 static bool
-vm_ic_hit_p(const struct iseq_inline_constant_cache_entry *ice, const VALUE *reg_ep)
+vm_ic_hit_p(const struct iseq_inline_constant_cache *ic, const VALUE *reg_ep)
 {
-    VM_ASSERT(IMEMO_TYPE_P(ice, imemo_constcache));
-    return vm_inlined_ic_hit_p(ice->flags, ice->value, ice->ic_cref, reg_ep);
+    return vm_inlined_ic_hit_p(vm_icc_flags(ic), vm_icc_value(ic), vm_icc_cref(ic), reg_ep);
 }
 
 // YJIT needs this function to never allocate and never raise
 bool
 rb_vm_ic_hit_p(IC ic, const VALUE *reg_ep)
 {
-    return ic->entry && vm_ic_hit_p(ic->entry, reg_ep);
+    return vm_icc_is_set(ic) && vm_ic_hit_p(ic, reg_ep);
 }
 
 static void
 vm_ic_update(const rb_iseq_t *iseq, IC ic, VALUE val, const VALUE *reg_ep, const VALUE *pc)
 {
+    vm_icc_reset(ic);
+
     if (ruby_vm_const_missing_count > 0) {
         ruby_vm_const_missing_count = 0;
-        ic->entry = NULL;
         return;
     }
 
-    struct iseq_inline_constant_cache_entry *ice = IMEMO_NEW(struct iseq_inline_constant_cache_entry, imemo_constcache, 0);
-    RB_OBJ_WRITE(ice, &ice->value, val);
-    ice->ic_cref = vm_get_const_key_cref(reg_ep);
-    if (rb_ractor_shareable_p(val)) ice->flags |= IMEMO_CONST_CACHE_SHAREABLE;
-    RB_OBJ_WRITE(iseq, &ic->entry, ice);
+    const rb_cref_t *cref = vm_get_const_key_cref(reg_ep);
+    if (cref) {
+        vm_icc_set_flag(ic, CONST_CACHE_HAS_EXT);
+        struct iseq_inline_constant_cache_entry *ice = IMEMO_NEW(struct iseq_inline_constant_cache_entry, imemo_constcache, 0);
+        RB_OBJ_WRITE(iseq, &ic->c.ext, ice);
+        RB_OBJ_WRITE(ic->c.ext, &ice->value, val);
+        ice->ic_cref = cref;
+    }
```

Comment:
So @ko1 made it always allocate an imemo for ractor data race reasons, that way cache fills are always an atomic pointer swap: e7fc353f044f9280222ca41b029b1368d2bf2fe3

The multiple pointer sets here doesn't look thread safe.

Reply:
Oh, that's good info thank you. I'll see if I can preserve the thread safety. I think with a bit of care it is possible.

Reply:
Hum, actually, not so sure, the problem is really that both `RACTOR_SHAREABLE` flag and the `VALUE` have to be set atomically, so short of having access to 128bit atomics, I don't see how it could be achieved 😢